core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1202,37 +1250,54 @@ private void deleteInternal(LockedInodePath inodePath, boolean replayed, long op
                     }
                   }
                 } else {
-                  if (!ufs.deleteDirectory(ufsUri, alluxio.underfs.options.DeleteOptions
-                      .defaults().setRecursive(true))) {
-                    failedToDelete = ufs.isDirectory(ufsUri);
-                    if (!failedToDelete) {
-                      LOG.warn("The directory to delete does not exist in ufs: {}", ufsUri);
+                  if (safeRecursiveUFSDeletes.containsKey(alluxioUriToDel)) {
+                    if (!ufs.deleteDirectory(ufsUri,
+                        alluxio.underfs.options.DeleteOptions.defaults().setRecursive(true))) {
+                      // TODO(adit): handle partial failures of recursive deletes
+                      failedToDelete = ufs.isDirectory(ufsUri);
+                      if (!failedToDelete) {
+                        LOG.warn("The directory to delete does not exist in ufs: {}", ufsUri);
+                      }
                     }
+                  } else {
+                    failedToDelete = true;
```

Comment:
If not safe for recursive => at least one new descendant in UFS => UFS directory cannot be deleted => directory inode cannot be deleted. Makes sense?

We will individually delete descendant inodes just not the unsafe directory itself.